File: tests/net/java/android/net/TcpKeepalivePacketDataTest.java
/*
 * Copyright (C) 2019 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package android.net;

import static org.junit.Assert.assertArrayEquals;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;

import android.net.SocketKeepalive.InvalidPacketException;
import android.net.TcpKeepalivePacketData.TcpSocketInfo;

import com.android.internal.util.TestUtils;

import libcore.net.InetAddressUtils;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.net.InetAddress;
import java.nio.ByteBuffer;

@RunWith(JUnit4.class)
public final class TcpKeepalivePacketDataTest {

    @Before
    public void setUp() {}

    @Test
    public void testV4TcpKeepalivePacket() {
        final InetAddress srcAddr = InetAddressUtils.parseNumericAddress("192.168.0.1");
        final InetAddress dstAddr = InetAddressUtils.parseNumericAddress("192.168.0.10");
        final int srcPort = 1234;
        final int dstPort = 4321;
        final int seq = 0x11111111;
        final int ack = 0x22222222;
        final int wnd = 8000;
        final int wndScale = 2;
        TcpKeepalivePacketData resultData = null;
        TcpSocketInfo testInfo = new TcpSocketInfo(
                srcAddr, srcPort, dstAddr, dstPort, seq, ack, wnd, wndScale);
        try {
            resultData = TcpKeepalivePacketData.tcpKeepalivePacket(testInfo);
        } catch (InvalidPacketException e) {
            fail("InvalidPacketException: " + e);
        }

        assertEquals(testInfo.srcAddress, resultData.srcAddress);
        assertEquals(testInfo.dstAddress, resultData.dstAddress);
        assertEquals(testInfo.srcPort, resultData.srcPort);
        assertEquals(testInfo.dstPort, resultData.dstPort);
        assertEquals(testInfo.seq, resultData.tcpSeq);
        assertEquals(testInfo.ack, resultData.tcpAck);
        assertEquals(testInfo.rcvWndScale, resultData.tcpWndScale);

        TestUtils.assertParcelingIsLossless(resultData, TcpKeepalivePacketData.CREATOR);

        final byte[] packet = resultData.getPacket();
        // IP version and TOS.
        ByteBuffer buf = ByteBuffer.wrap(packet);
        assertEquals(buf.getShort(), 0x4500);
        // Source IP address.
        byte[] ip = new byte[4];
        buf = ByteBuffer.wrap(packet, 12, 4);
        buf.get(ip);
        assertArrayEquals(ip, srcAddr.getAddress());
        // Destination IP address.
        buf = ByteBuffer.wrap(packet, 16, 4);
        buf.get(ip);
        assertArrayEquals(ip, dstAddr.getAddress());

        buf = ByteBuffer.wrap(packet, 20, 12);
        // Source port.
        assertEquals(buf.getShort(), srcPort);
        // Destination port.
        assertEquals(buf.getShort(), dstPort);
        // Sequence number.
        assertEquals(buf.getInt(), seq);
        // Ack.
        assertEquals(buf.getInt(), ack);
        // Window size.
        buf = ByteBuffer.wrap(packet, 34, 2);
        assertEquals(buf.getShort(), wnd >> wndScale);
    }

    //TODO: add ipv6 test when ipv6 supported

    @Test
    public void testParcel() throws Exception {
        final InetAddress srcAddr = InetAddresses.parseNumericAddress("192.168.0.1");
        final InetAddress dstAddr = InetAddresses.parseNumericAddress("192.168.0.10");
        final int srcPort = 1234;
        final int dstPort = 4321;
        final int sequence = 0x11111111;
        final int ack = 0x22222222;
        final int wnd = 48_000;
        final int wndScale = 2;
        TcpKeepalivePacketData testData = null;
        TcpKeepalivePacketDataParcelable resultData = null;
        TcpSocketInfo testInfo = new TcpSocketInfo(
                srcAddr, srcPort, dstAddr, dstPort, sequence, ack, wnd, wndScale);
        testData = TcpKeepalivePacketData.tcpKeepalivePacket(testInfo);
        resultData = testData.toStableParcelable();
        assertArrayEquals(resultData.srcAddress, srcAddr.getAddress());
        assertArrayEquals(resultData.dstAddress, dstAddr.getAddress());
        assertEquals(resultData.srcPort, srcPort);
        assertEquals(resultData.dstPort, dstPort);
        assertEquals(resultData.seq, sequence);
        assertEquals(resultData.ack, ack);
    }
}

File: tests/net/common/java/android/net/TcpKeepalivePacketDataTest.kt
/*
 * Copyright (C) 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package android.net

import android.net.InetAddresses.parseNumericAddress
import android.os.Build
import com.android.testutils.DevSdkIgnoreRule
import com.android.testutils.DevSdkIgnoreRunner
import com.android.testutils.assertFieldCountEquals
import com.android.testutils.assertParcelSane
import org.junit.Test
import org.junit.runner.RunWith
import java.net.InetAddress
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals
import kotlin.test.assertTrue

@RunWith(DevSdkIgnoreRunner::class)
@DevSdkIgnoreRule.IgnoreUpTo(Build.VERSION_CODES.R) // TcpKeepalivePacketData added to SDK in S
class TcpKeepalivePacketDataTest {
    private fun makeData(
        srcAddress: InetAddress = parseNumericAddress("192.0.2.123"),
        srcPort: Int = 1234,
        dstAddress: InetAddress = parseNumericAddress("192.0.2.231"),
        dstPort: Int = 4321,
        data: ByteArray = byteArrayOf(1, 2, 3),
        tcpSeq: Int = 135,
        tcpAck: Int = 246,
        tcpWnd: Int = 1234,
        tcpWndScale: Int = 2,
        ipTos: Int = 0x12,
        ipTtl: Int = 10
    ) = TcpKeepalivePacketData(srcAddress, srcPort, dstAddress, dstPort, data, tcpSeq, tcpAck,
            tcpWnd, tcpWndScale, ipTos, ipTtl)

    @Test
    fun testEquals() {
        val data1 = makeData()
        val data2 = makeData()
        assertEquals(data1, data2)
        assertEquals(data1.hashCode(), data2.hashCode())
    }

    @Test
    fun testNotEquals() {
        assertNotEquals(makeData(srcAddress = parseNumericAddress("192.0.2.124")), makeData())
        assertNotEquals(makeData(srcPort = 1235), makeData())
        assertNotEquals(makeData(dstAddress = parseNumericAddress("192.0.2.232")), makeData())
        assertNotEquals(makeData(dstPort = 4322), makeData())
        // .equals does not test .packet, as it should be generated from the other fields
        assertNotEquals(makeData(tcpSeq = 136), makeData())
        assertNotEquals(makeData(tcpAck = 247), makeData())
        assertNotEquals(makeData(tcpWnd = 1235), makeData())
        assertNotEquals(makeData(tcpWndScale = 3), makeData())
        assertNotEquals(makeData(ipTos = 0x14), makeData())
        assertNotEquals(makeData(ipTtl = 11), makeData())

        // Update above assertions if field is added
        assertFieldCountEquals(5, KeepalivePacketData::class.java)
        assertFieldCountEquals(6, TcpKeepalivePacketData::class.java)
    }

    @Test
    fun testParcelUnparcel() {
        assertParcelSane(makeData(), fieldCount = 6) { a, b ->
            // .equals() does not verify .packet
            a == b && a.packet contentEquals b.packet
        }
    }

    @Test
    fun testToString() {
        val data = makeData()
        val str = data.toString()

        assertTrue(str.contains(data.srcAddress.hostAddress))
        assertTrue(str.contains(data.srcPort.toString()))
        assertTrue(str.contains(data.dstAddress.hostAddress))
        assertTrue(str.contains(data.dstPort.toString()))
        // .packet not included in toString()
        assertTrue(str.contains(data.getTcpSeq().toString()))
        assertTrue(str.contains(data.getTcpAck().toString()))
        assertTrue(str.contains(data.getTcpWindow().toString()))
        assertTrue(str.contains(data.getTcpWindowScale().toString()))
        assertTrue(str.contains(data.getIpTos().toString()))
        assertTrue(str.contains(data.getIpTtl().toString()))

        // Update above assertions if field is added
        assertFieldCountEquals(5, KeepalivePacketData::class.java)
        assertFieldCountEquals(6, TcpKeepalivePacketData::class.java)
    }
}